core/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -704,7 +705,74 @@ private void cleanupRestoreState(ClusterChangedEvent event) {
+                    }
+                });
+            }
+        }
+    }
+
+    private ClusterState resyncRestoreInProgressWithRoutingTable(ClusterState currentState) {
+        RoutingTable routingTable = currentState.routingTable();
+        RestoreInProgress restoreInProgress = currentState.custom(RestoreInProgress.TYPE);
+        if (restoreInProgress != null) {
+            final List<RestoreInProgress.Entry> entries = new ArrayList<>();
+            for (RestoreInProgress.Entry entry : restoreInProgress.entries()) {
+                Snapshot snapshot = entry.snapshot();
+                ImmutableOpenMap.Builder<ShardId, ShardRestoreStatus> shardsBuilder = ImmutableOpenMap.builder(entry.shards());
+                for (ObjectObjectCursor<ShardId, ShardRestoreStatus> restoreEntry : entry.shards()) {
+                    ShardId shardId = restoreEntry.key;
+                    IndexShardRoutingTable indexShardRoutingTable = routingTable.shardRoutingTableOrNull(shardId);
+                    if (indexShardRoutingTable == null) {
+                        shardsBuilder.put(shardId, new ShardRestoreStatus(null, RestoreInProgress.State.FAILURE, "index was deleted"));
+                    } else {
+                        if (restoreEntry.value.state().completed() == false) {
+                            // must be INIT (restore state for ShardRestoreStatus is never STARTED)
+                            assert restoreEntry.value.state() == RestoreInProgress.State.INIT;
+                            ShardRouting primaryShard = indexShardRoutingTable.primaryShard();
+                            if (primaryShard.active()) {
+                                // assume shard was started after successful restore from snapshot
+                                shardsBuilder.put(shardId,
+                                    new ShardRestoreStatus(primaryShard.currentNodeId(), RestoreInProgress.State.SUCCESS));
+                            } else {
+                                assert primaryShard.unassigned() || primaryShard.initializing();
+                                if (primaryShard.recoverySource().getType() == RecoverySource.Type.SNAPSHOT &&
+                                    ((SnapshotRecoverySource) primaryShard.recoverySource()).snapshot().equals(snapshot)) {
+                                    // if the primary is unassigned we assume that the restore wasn't started yet and try the restore again
+                                    // once the primary is assigned to a node.
+                                    // if the primary is initializing we assume that the restore is in progress.
+                                } else {
+                                    shardsBuilder.put(shardId, new ShardRestoreStatus(null, RestoreInProgress.State.FAILURE,
+                                        "recovery source type changed from snapshot to " + primaryShard.recoverySource()));
```

Comment:
Hmm, I wonder if we can come up with a higher level description here. Can you think of any simple scenario when this could happen (besides somebody deleting an index while snapshot was running and creating a new one)?